### PR TITLE
dev/cloud-native#12: Define the BASE URL via an environment variable at run time

### PR DESCRIPTION
**Pull request by tiotsop01**
Overview
----------------------------------------
Making it easy to provide the absolute system path at runtime from the environment in CiviCRM rather than hardcoded in a file

Before
----------------------------------------
CiviCRM was outputting URLs when it could/should be outputting paths. This is important because URLs (absolute ones) get in the way of reverse proxying which is something cloud providers typically want to do.

After
----------------------------------------
Canonical hostname and port are now defined and stored as environmental variables.

### CODE REVIEW

**Comment by kcristiano** [comment]
@tiotsop01 Thanks for the patch.  Is there an issue on Gitlab for this?   Can you also comment on the testing you've done and for which CMS's that you have done this?

**Comment by michaelmcandrew** [comment]
@kcristiano FYI this is a work in progress and not ready for review. I have added a label to mark it as such.

The only reason we have created a PR at this stage is because we want to run

`civibuild create dmaster --patch https://github.com/civicrm/civicrm-core/pull/12263`

**Comment by tiotsop01** [comment]
@michaelmcandrew I have updated this PR. Awaiting review

**Review by michaelmcandrew — COMMENTED** [review]
@tiotsop01 - did you do any testing of this on WordPress? Probably worth doing a quick test and reporting back what happens. We don't have to fix it necessarily but would be good to know how it behaves.

**Comment by michaelmcandrew** [comment]
@tiotsop01 and @totten I'm looking at https://github.com/civicrm/civicrm-buildkit/tree/master/tests/phpunit and thinking about how we could create a buildkit test for this.

I suspect that creating a test that uses Civi\Civibuild\ProcessUtil to do something like the following:

`CIVICRM_UF_BASEURL=http://proxied-url.example.org civibuild create dmaster`

and then

`CIVICRM_UF_BASEURL=http://proxied-url.example.org  cv ev "echo CIVICRM_UF_BASEURL"`

from the appropriate directory and checking that the output = http://proxied-url.example.org is a good enough test.

Any thoughts?

Hopefully \Symfony\Component\Process\Process (on which Civi\Civibuild\ProcessUtil is built) allows one to pass environment variables as part of the command.

File: templates/CRM/common/civicrm.settings.php.template [hunk 1]
```diff
@@ -214,11 +214,26 @@ if (!defined('CIVICRM_TEMPLATE_COMPILEDIR')) {
...
-  define( 'CIVICRM_UF_BASEURL'      , '%%baseURL%%');
+  if (getenv('CIVICRM_UF_BASEURL')) {
+    define('CIVICRM_UF_BASEURL', getenv('CIVICRM_UF_BASEURL'));
+  }
+  else {
+    define('CIVICRM_UF_BASEURL', '%%baseURL%%');
+  }
 }
 
+//if (!defined('CIVICRM_UF_BASEURL')) {
```
[the remark this answers is not in the document]
Comment (tiotsop01) [reply templates/CRM/common/civicrm.settings.php.template]: Yes, it is.

[the remark this answers is not in the document]
Comment (tiotsop01) [reply templates/CRM/common/civicrm.settings.php.template]: @michaelmcandrew This is how the test on WordPress looks.  
 `civibuild create wpmaster --url http://www.civi-example.com:8080`
wpmaster

[the remark this answers is not in the document]
Comment (tiotsop01) [reply templates/CRM/common/civicrm.settings.php.template]: I have also removed the unused snippet

[the remark this answers is not in the document]
Comment (michaelmcandrew) [reply templates/CRM/common/civicrm.settings.php.template]: @tiotsop01 - cool - that looks good.

**Comment by eileenmcnaughton** [comment]
@michaelmcandrew @tiotsop01 this has kinda stalled. It looks like a safe patch since you'd need to deliberately set the env var for it to have any effect. But, the discussion / consensus is lacking. I'm inclined to close it & resolve that discussion in gitlab now before re-opening & merging - unless either of you are keen to push on with this now?

**Comment by michaelmcandrew** [comment]
since we haven't heard from @tiotsop01, lets close. we can re-open if / when he wants to push it forward

**Comment by eileenmcnaughton** [comment]
done

**Pull request closed without merging**